net/connstats/stats.go

```diff
@@ -135,19 +135,19 @@ func (s *Statistics) updateVirtual(b []byte, receive bool) {
 // The src is always a Tailscale IP address, representing some remote peer.
 // The dst is a remote IP address and port that corresponds
 // with some physical peer backing the Tailscale IP address.
-func (s *Statistics) UpdateTxPhysical(src netip.Addr, dst netip.AddrPort, n int) {
-	s.updatePhysical(src, dst, n, false)
+func (s *Statistics) UpdateTxPhysical(src netip.Addr, dst netip.AddrPort, packets, bytes int) {
+	s.updatePhysical(src, dst, packets, bytes, false)
 }
 
 // UpdateRxPhysical updates the counters for a received wireguard packet.
 // The src is always a Tailscale IP address, representing some remote peer.
 // The dst is a remote IP address and port that corresponds
 // with some physical peer backing the Tailscale IP address.
-func (s *Statistics) UpdateRxPhysical(src netip.Addr, dst netip.AddrPort, n int) {
-	s.updatePhysical(src, dst, n, true)
+func (s *Statistics) UpdateRxPhysical(src netip.Addr, dst netip.AddrPort, packets, bytes int) {
+	s.updatePhysical(src, dst, packets, bytes, true)
 }
 
-func (s *Statistics) updatePhysical(src netip.Addr, dst netip.AddrPort, n int, receive bool) {
+func (s *Statistics) updatePhysical(src netip.Addr, dst netip.AddrPort, packets, bytes int, receive bool) {
 	conn := netlogtype.Connection{Src: netip.AddrPortFrom(src, 0), Dst: dst}
 
 	s.mu.Lock()
```

Comment:
as this is now planned to be on all the time, we should work to remove this lock from the data path as well

Reply:
It isnt really planned to be on all the time, the metrics I'm working on doesnt get the lock and sits next to this. I happened to discover that the packets were not counted when I was comparing the metrics to the stats, because we wanted to ensure they were always the same.

Reply:
The quite messy draft pr https://github.com/tailscale/tailscale/pull/13309 is the metrics parts, this, tests etc. I am currently breaking it up into multiple PRs, tho some of them will still be sizable.

Reply:
I see, the doc referenced from the bug this PR references isn't clear, but seems to imply metrics are on all the time.

We're sensitive in the data path, additional mutex's often take >= 5% from our peak throughput (and more on non-offload platforms), as packets are frequent and the mutex's are too coarse to avoid contention (particularly when allocs occur under the lock). fwiw, this shows up on full flows, and won't show up on microbenchmarks, because we use more than ncpu live goroutines under full data path load and the higher contention costs come from entering/exiting the scheduler/switching goroutines.

Are you adding new metrics? What will be the data format and allocation patterns for those?

Reply:
oh, I re-read my own comment and I am quite unclear, apologies.

The change in this pr, the stats are not going to be on all the time, they will remain like now, this change is just so I can compare them with metrics.

Metrics are intended to be on all the time, they are based on expvar + our Multilabelmap, I've just opened https://github.com/tailscale/tailscale/pull/13550 which shows how they are implemented for _other_ parts than the wgengine (but tswrap for example). 

The current implementation for wgengine in #13309 (which I'll split out to a separate PR later) does not use or introduce any new mutexs, except for the one that exists inside the individual metric/counter. So it should not hold the coarse mutex. If you have time, please have a look at #13550, and I will make sure to include you and @jwhited on the PR for the metrics in wgengine.

Reply:
Do we have any data for the how much of a bottleneck this mutex is?

I have an old prototype for a mostly lock-free map of connection counts. 
https://github.com/tailscale/tailscale/commit/b63618452c863265c4a22d45f67a6d0da6e33aa8

IIRC, the performance of this was much better, but prototype could use more polishing. I was also hesitant to add the more complex optimization without proof that it was solving a real problem.

Reply:
It sounds like performance of the stats is a bit of a separate discussion from this, but performance of the metrics, to be added in a following PR is not, I will make sure to include you in the review for that.